### PR TITLE
[WIP]MIG-64: Mixed variation migration

### PR DESCRIPTION
**Pull request by momoss**
**Description**

**Definition Of Done**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/MixedVariationProductMigrator.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+        foreach ($variantGroupCombination->getGroups() as $variantGroupCode) {
+            $rootProductModel = $this->productModelBuilder->buildRootProductModel($variantGroupCode, $familyVariant, $pim);
+            $rootProductModel = $this->productModelRepository->persist($rootProductModel, $pim);
+
+            $productsHavingVariants = $mixedVariation->getVariantGroupProductsHavingVariants($variantGroupCode);
+
+            foreach ($productsHavingVariants as $parentProduct) {
+                $subProductModel = $this->productModelBuilder->buildSubProductModel($rootProductModel, $parentProduct, $familyVariant, $pim);
+                $subProductModel = $this->productModelSaver->save($subProductModel, $pim);
```
Comment (LaurentPetard) [inline src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/MixedVariationProductMigrator.php]: That will be replaced by `productModelRepository->persist` in a next PR where the repository will use the API to persist. I can't use the import without doing useless work to re-format the attributes values.

File: src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/MixedVariationMigrator.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    public function migrate(SourcePim $sourcePim, DestinationPim $destinationPim): void
+    {
+        $variantGroupCombinations = $this->variantGroupCombinationRepository->findAll($destinationPim);
+        $hasInvalidMixedVariations = false;
+
+        foreach ($variantGroupCombinations as $variantGroupCombination) {
+            $mixedVariation = $this->mixedVariationBuilder->buildFromVariantGroupCombination($variantGroupCombination, $destinationPim);
+
+            if (!$mixedVariation instanceof MixedVariation) {
```
Comment (LaurentPetard) [inline src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/MixedVariationMigrator.php]: null === $mixedVariation

File: src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/FamilyVariantLabelBuilder.php [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    /**
+     * Family variant labels = Labels of the family + Labels of the variant axes attributes + Labels of the inner variation type, separated by a space.
+     */
+    public function build(VariantGroupCombination $variantGroupCombination,  InnerVariationType $innerVariationType, DestinationPim $pim)
+    {
+        $familyVariantLabels = $variantGroupCombination->getFamily()->getStandardData()['labels'];
+
+        foreach ($variantGroupCombination->getAxes() as $axe) {
```
Comment (LaurentPetard) [inline src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/FamilyVariantLabelBuilder.php]: as axis

File: src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/MixedVariationProductMigrator.php [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                    $variantGroupCombination->getFamily(),
+                    $innerVariationType->getVariationFamily(),
+                    $pim
+                );
+
+                $this->productRepository->delete($parentProduct->getIdentifier(), $pim);
+            }
+
+            // To transform into variants the remaining products that would not have variations via the IVB.
+            $this->variantGroupProductVariantTransformer->transformFromProductModel($rootProductModel, $familyVariant, $pim);
```
Comment (LaurentPetard) [inline src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/MixedVariationProductMigrator.php]: Nope. all the products with mixed variation must be migrated with 2 levels.

Comment (LaurentPetard) [reply src/Domain/MigrationStep/s150_ProductVariationMigration/MixedVariation/MixedVariationProductMigrator.php]: Fixed

File: src/Domain/Command/Api/CreateProductModelCommand.php [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ */
+class CreateProductModelCommand implements ApiCommand
+{
+    /** @var string */
+    private $code;
+
+    /** @var array */
+    private $data;
+
+    public function __construct($code, array $data)
```
Comment (anaelChardan) [inline src/Domain/Command/Api/CreateProductModelCommand.php]: string code

**Comment by anaelChardan** [comment]
It misses the description on your PR ;)

File: src/Domain/MigrationStep/s150_ProductVariationMigration/Entity/ProductModel.php [hunk 6]
```diff
@@ -27,18 +27,23 @@ class ProductModel
...
+    private $parentIdentifier;
+
     public function __construct(
         ?int $id,
         string $identifier,
         string $familyVariantCode,
         array $categories,
-        array $attributeValues
+        array $attributeValues,
+        string $parentIdentifier = null
```
Comment (ahocquard) [inline src/Domain/MigrationStep/s150_ProductVariationMigration/Entity/ProductModel.php]: `?string` or default value ?

File: src/Domain/MigrationStep/s150_ProductVariationMigration/ProductRepository.php [hunk 7]
```diff
@@ -64,12 +65,43 @@ public function findAllNotMigratedProductVariants(Pim $pim): \Iterator
...
+SQL;
+
+        $results = $this->console->execute(new MySqlQueryCommand(sprintf($query,$groupCode)), $pim)->getOutput();
+
+        $products = [];
+        foreach ($results as $productData) {
+            $products[] = new Product(
+                (int) $productData['id'],
+                $productData['identifier'],
+                (int) $productData['family_id'],
```
[the remark this answers is not in the document]
Comment (LaurentPetard) [reply src/Domain/MigrationStep/s150_ProductVariationMigration/ProductRepository.php]: Can not be null. The family is mandatory in 2.0

File: src/Domain/MigrationStep/s150_ProductVariationMigration/Entity/ProductModel.php (hunk 6 above)
Comment (LaurentPetard) [reply src/Domain/MigrationStep/s150_ProductVariationMigration/Entity/ProductModel.php]: default value

File: src/Domain/MigrationStep/s150_ProductVariationMigration/InnerVariation/InnerVariationTypeRepository.php [hunk 8]
```diff
@@ -104,6 +135,34 @@ public function getLabel(InnerVariationType $innerVariationType, string $locale,
...
+            return null;
+        }
+
+        $innerVariationTypeData = $innerVariationTypeData[0];
+        $innerVariationTypeId = (int) $innerVariationTypeData['id'];
+
+        return new InnerVariationType(
+            $innerVariationTypeId,
+            $innerVariationTypeData['code'],
+            $this->familyRepository->findById((int) $innerVariationTypeData['variation_family_id'], $pim),
```
[the remark this answers is not in the document]
Comment (LaurentPetard) [reply src/Domain/MigrationStep/s150_ProductVariationMigration/InnerVariation/InnerVariationTypeRepository.php]: fixed

**Pull request merged**